tensorboard/webapp/widgets/histogram/histogram_v2_component.ts

```diff
@@ -55,83 +77,178 @@ export class HistogramV2Component implements OnChanges {
 
   @Input() data!: HistogramData;
 
-  private yScale?:
-    | d3.ScaleTime<number, number>
-    | d3.ScaleLinear<number, number>;
+  readonly HistogramMode = HistogramMode;
 
-  constructor(private readonly host: ElementRef) {}
+  private layout: Layout = {
+    histogramHeight: 1,
+    contentClientRect: {height: 1, width: 1},
+  };
+  private scales: Scales;
+  private domInitialized = false;
+
+  constructor(private readonly changeDetector: ChangeDetectorRef) {
+    // `data` may not be available at the constructor time. Since we recalculate
+    // the scales on the `ngAfterViewInit`, let's just initialize `scales` with
+    // their default values.
+    this.scales = this.updateScales(this.layout, []);
+  }
 
   ngOnChanges() {
     if (this.content) {
-      this.ngAfterViewInit();
+      this.updateChart();
     }
   }
 
   ngAfterViewInit() {
+    this.domInitialized = true;
+    this.updateClientRects();
     this.updateChart();
+    this.changeDetector.detectChanges();
+  }
+
+  getHistogramPath(datum: HistogramDatum): string {
+    if (!this.domInitialized || !datum.bins.length) return '';
+    const xScale = this.scales.binScale;
+    const yScale = this.scales.countScale;
+
+    const firstBin = datum.bins[0];
+    const lastBin = datum.bins.slice(-1)[0];
+    const pathBuilder = [
+      `M${xScale(firstBin.x + firstBin.dx * 0.5)},${yScale(0)}`,
+    ];
+
+    for (const bin of datum.bins) {
+      pathBuilder.push(`L${xScale(bin.x + bin.dx * 0.5)},${yScale(bin.y)}`);
+    }
+
+    pathBuilder.push(`L${xScale(lastBin.x + lastBin.dx * 0.5)},${yScale(0)}`);
+    pathBuilder.push('Z');
+    return pathBuilder.join('');
+  }
+
+  trackByWallTime(datum: HistogramDatum): number {
+    return datum.wallTime;
+  }
+
+  // translates container for histogram so we can have more sensible coordinate
+  // system for reasoning with the coordinate system of a histogram.
+  getGroupTransform(datum: HistogramDatum): string {
+    if (!this.domInitialized || this.mode === HistogramMode.OVERLAY) return '';
```

Comment:
I'm not sure why some methods like `getGroupTransform`, `getHistogramPath` check `this.domInitialized` while others like `getHistogramFill`, `getGridTickYLocs` do not.  All 4 are used in the template.  How did we decide which ones require the `domInitialized` check?